docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,134 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
```

Comment:
Assuming you want to limit the number of connections in a trusted environment the IP address is not the best option I'd like to have to rate limit the connection creation rate. Instead as an ops user I'd like to be able to limit this per SASL user (additionally).

**Here is why:**
We had issues with connection creation rate in my previous company. Similarly how we do it at Redpanda SaaS, we had Kafka running on one cluster and peered several other networks that could access the kafka clusters. We had >500 microservices from >70 teams using our clusters. We only knew that any service is creating too many connections - but we had no idea what service or team it is. We had a hard time figuring this out because we were using NAT and IP masquerading. There were no logs emitted what SASL user is creating these many connections which would have helped us (even though I understand that the rate limit may happen before authentication)

Reply:
This RFC will for sure help but it might not be enough for clients behind NATs, that's why I recently added [KIP-713](https://cwiki.apache.org/confluence/display/KAFKA/KIP-731%3A+Record+Rate+Limiting+for+Kafka+Connect) to the list we have here: https://docs.google.com/document/d/1NQTfklE4SQD2USf8diclZq1v4dJ-fs4C06buor_ZkSc. 

This above is to say that we should still move forward with this RFC and schedule work to look into KIP-731.

Reply:
@weeco the discussion becomes a bit nuanced when talking about per-user limits, as we cannot know what user is behind a connection until we have accepted it and received some bytes.

I think there's a much broader topic of how to manage traffic in general & also how to get visibility into "who's hammering my system" at a username/IP/topic level, which we should address but not here.